fem/bilinearform_ext.cpp

```diff
@@ -255,7 +255,8 @@ PABilinearFormExtension::PABilinearFormExtension(BilinearForm *form)
 void PABilinearFormExtension::SetupRestrictionOperators(const L2FaceValues m)
 {
    if ( Device::Allows(Backend::CEED_MASK) ) { return; }
-   ElementDofOrdering ordering = UsesTensorBasis(*a->FESpace())?
+   ElementDofOrdering ordering = UsesTensorBasis(*a->FESpace()) &&
+                                 use_tensor_basis ?
```

Comment:
I did not think about that aspect. There would need to be some extra checks involved.

Reply:
One solution is to give `BilinearFormIntegrator`  extra members indicating which type of partial assembly it supports, and the extension would have enough info to make sure it doesn't mix up the orderings.

Reply:
I think it would be better to use the `ordering` from the `BilinearFormExtension` in `GetDofToQuad(*IntRule, DofToQuad::FULL)`, something like `GetDofToQuad(*IntRule, DofToQuad::FULL, ordering)`. Like this the non-tensor implementation and tensor implementation of different integrators can both work with a lexicographic restriction without creating any silent bugs.

Reply:
I will look into implementing this idea.

Reply:
This should be fairly straightforward using the `GetDofMap` method, it gives you the mapping between native and lexicographic indices. Then you just have to use it in `ScalarFiniteElement::GetDofToQuad` and you should be good.
@v-dobrev was also proposing to use a third enum value for `DofToQuad::Mode`, something like `LEXICOGRAPHIC_FULL`.

Reply:
I actually just pushed a change to my PR using the `GetDofToQuad(*IntRule, DofToQuad::FULL, ordering)` overload idea. I made it a virtual function for `FiniteElement` and overrode it in `NodalTensorFiniteElement` to support simultaneous `LEXICOGRAPHIC` and `FULL` ordering. It might be equivalent to using `GetDofMap` because I use the `dof_map` member to undo the native mapping. As I was doing this, I had also thought of adding a third enum value like @v-dobrev was thinking. But I was able to revert my proposed changes to `BilinearFormExtension` and `BilinearFormExtension` at the expense of changes to `FiniteElement` that should be more hidden from the user.

Reply:
Note I only did it for the shape funtion gradients so far, I need to do it for the basis values as well.